File: 2024-1_3P/2024-1_TPs/2024-1_TP1/palindromoR.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

bool palindromo(char* s, int i)
{
    int length = strlen(s);   
    if( i >= length/2)
    {
        return true; // chegou no meio da string sem diferenças, então é um palindromo
    } // end if

    int j = length - 1 - i;

    if(s[i] != s[j])
    {
        return false;
    } // end if 
   
   return palindromo(s, i + 1);
} // end palindromo

int main(void)
{
    // definir dados 
    char s[1000];
    // obter dados e aplicar função auxiliar
     while(true)
    {
        scanf ( "%[^\n]", s); 
        
        if(strcmp(s, "FIM") == 0)
        {
            break; // Saindo do loop se a entrada for "FIM"
        } // end if 

        // Limpar o buffer de entrada
        while (getchar() != '\n');

        // Verificando se a entrada é um palíndromo
        bool result = palindromo(s,0);
        
        // Imprimindo resultado
        if(result)
        {
            printf("SIM\n");
        }
        else
        {
            printf("NAO\n");
        } // end if 
    } // end while
    
    return 0;
} // end main

File: 2024-1_3P/2024-1_TP1/palindromoR.java
public class palindromoR
{
    public static boolean palindromo(String s, int i )
    {
        int length = s.length();
        if(i >= length) // se chegar até metade da string sem diferenças é palindromo
        {
            return true; 
        } // end if 
        int j = length - 1 - i;
        if(s.charAt(i) != s.charAt(j))
        {
            return false;
        } // end if
        return palindromo(s, i + 1);
    } // end palindromo

    public static void main(String[] args)
    {
        String s; boolean result; 
        while(true)
        {
        // definir dados 
        s = MyIO.readLine();

        if(s.equals("FIM"))
        {
            break;
        } // end if

        // aplicar funções auxiliares
        result = palindromo(s,0);

        if(result == true)
        {
            MyIO.println("SIM");
        }
        else
        {
            MyIO.println("NAO");
        } // end if 
        } // 
    } // end main( )
} // end palindromoR
